### DISCUSSION TITLE
Scripting and webserver

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#17918] [Q&A] [answered]
Would like to compile with #undef USE_WEBSERVER to save 8kb RAM so I can use #define USE_GPS and #define USE_SCRIPTING without being near the edge of RAM space (so I can log GPS to flash and not use FLOG, which eats a ton of RAM and only writes every 5 minutes). However, the build fails with 
```
/Users/andrewlotze/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o:(.text._Z6Xdrv10j+0x28): undefined reference to `set_callbacks()'
/Users/andrewlotze/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o:(.text._Z6Xdrv10j+0x2c): undefined reference to `script_set_web_pages()'
/Users/andrewlotze/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o: in function `Xdrv10(unsigned int)':
tasmota.ino.cpp:(.text._Z6Xdrv10j+0x243): undefined reference to `set_callbacks()'
/Users/andrewlotze/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: tasmota.ino.cpp:(.text._Z6Xdrv10j+0x255): undefined reference to `script_set_web_pages()'
collect2: error: ld returned 1 exit status
*** [.pio/build/tasmota/firmware.elf] Error 1
```
@gemu2015 Is this something fixable, or is scripting too integrated into the GUI?

Another question about writing to flash. If i'm writing just a few bytes each time to a ufilesys file, is the system erasing and re-writing a 4096-byte block each time? Thus I should buffer it in RAM until I can write an entire block so I don't make 341 flash writes onto the same block?

**Comment by gemu2015** [comment, marked as answer]
i honestly never checked with #undef USE_WEBSERVER.
i fixed the compile error now easily however i never tested this use case.
there is no way to transfer a script via MQTT. (the fixed scripter.ino is in my fork)

according little file system, i guess it only writes if you close the file or a sector is filled but i am not sure about that.
so probably keep the file open and only close and reopen  it on larger intervals

**Reply by sfromis** [reply]
As I suggested on Discord, masochists can have one binary including web server to update the script, and then switch to another binary without web server, to run the script.

**Reply by pkkrusty** [reply]
Makes sense. 

Thanks for the fix, I'll pull from your fork and see how it goes. sfromis had the good idea to get my script in with a webserver build, then OTA update to a non-webserver build. Convoluted, but should work.

Next problem is how I convert lat lon that I get from the GPS driver to hex string so I can write into a 12-byte block. Not sure if variables are float or int_8 or _16 or _32. I only need 5 decimal place precision, so 7 digits total, plus decimal (but that can be done in post-processing). 4 bytes each for lat, lon and epoch.

**Reply by pkkrusty** [reply]
When I use the following:
```
>S
str=""
if lat>0 {
latnum=int(lat*100000)
lonnum=abs(int(lon*100000))
; hx() converts a 16-bit number to hex according to docs
str+=hx((epoch>>16)&0xFFFF)
str+=hx(epoch&0xFFFF)
str+=hx(latnum>>16)
str+=hx(latnum&0xFFFF)
;str+=hx(lonnum>>16)
;str+=hx(lonnum&0xFFFF)
}
print %str%,%5lat%,%0latnum%,%5lon%,%0lonnum%,%0epoch%,%0heap%
```
I get 
```
16:06:12.069 000007b900001af0000000160000028a,14.42442,1442442,-16.99308,1699309,129571568,18832
16:06:13.075 000007b900001af8000000160000028a,14.42442,1442442,-16.99308,1699309,129571576,18624
16:06:14.036 000007b900001af8000000160000028a,14.42442,1442442,-16.99308,1699309,129571576,18832
16:06:15.036 000007b900001af8000000160000028a,14.42443,1442442,-16.99309,1699309,129571576,18832
16:06:16.041 000007b900001af8000000160000028a,14.42443,1442442,-16.99309,1699310,129571576,18832
16:06:17.036 000007b900001af8000000160000028b,14.42444,1442443,-16.99310,1699311,129571576,18624
```
Instead of what I would expect 07b91af80016028b in the string.

**Reply by gemu2015** [reply]
sorry, error in docs, it converts to 32 bit hex

**Reply by pkkrusty** [reply]
Noting that variable `epoch` is only updating once every 8 seconds. Is this intentional or something weird with my build?

```
23:05:41.165 epoch 129596744.00
23:05:42.129 epoch 129596744.00
23:05:43.126 epoch 129596744.00
23:05:44.126 epoch 129596744.00
23:05:45.126 epoch 129596744.00
23:05:46.126 epoch 129596744.00
23:05:47.126 epoch 129596744.00
23:05:48.126 epoch 129596752.00
23:05:49.125 epoch 129596752.00
23:05:50.126 epoch 129596752.00
23:05:51.126 epoch 129596752.00
23:05:52.126 epoch 129596752.00
23:05:53.126 epoch 129596752.00
23:05:54.126 epoch 129596752.00
23:05:55.125 epoch 129596752.00
23:05:56.126 epoch 129596752.00
23:05:56.251 WIF: Checking connection...
23:05:57.126 epoch 129596760.00
23:05:58.126 epoch 129596760.00
23:05:59.126 epoch 129596760.00
23:06:00.126 epoch 129596760.00
23:06:01.125 epoch 129596760.00
23:06:02.126 epoch 129596760.00
23:06:03.126 epoch 129596760.00
23:06:04.126 epoch 129596768.00
23:06:05.126 epoch 129596768.00
23:06:06.126 epoch 129596768.00
23:06:07.126 epoch 129596768.00
23:06:08.126 epoch 129596768.00
23:06:09.126 epoch 129596768.00
23:06:10.126 epoch 129596768.00
23:06:11.126 epoch 129596768.00
23:06:12.126 epoch 129596768.00
23:06:13.126 epoch 129596776.00
23:06:14.126 epoch 129596776.00
23:06:15.126 epoch 129596776.00
23:06:16.126 epoch 129596776.00
23:06:16.251 WIF: Checking connection...
23:06:17.126 epoch 129596776.00
23:06:18.126 epoch 129596776.00
```

**Reply by barbudor** [reply]
Variables are float so only 23 bits mantissa

**Reply by pkkrusty** [reply]
Next issue: If I create a hex string, and then write that to a file, I'm wasting 50% space, since each character is 8-bit and I only want 4-bit. Simple solution that you can see? 

Maybe fwb() function?
```
cnt=sl(str)
for cntr 1 cnt 2
subst=sb(str cntr 2)
dcml=hd(subst)
res=fwb(dcml fr)
next
```
Not sure if I need to start at position 0 or 1 for sb() function.

**Reply by gemu2015** [reply]
like @barbudor  said, the 8 second update is simply due to the limit of float. 

i assume you want to write to a flash file on a 4MB device?
so i would 1. choose another linker file to get 3 MB of file system
and 2. would write binary data (if you have fixed record sizes) = fwb()

**Reply by pkkrusty** [reply]
So epoch is not really "seconds since 2019-01-01", but 23-bit float approximation of that? `epoffs` variable might work, but are there any arguments to pass for that?

**Reply by gemu2015** [reply]
redefine epoch offset currently at this value
#define EPOCH_OFFSET 1546300800
and you will have about 2 years of second accurate values

you may also write this value `epoffs=val`

**Reply by barbudor** [reply]
Epoch is internally counted as a 32 bit integer with second accuracy from Jan 1970 as the standard say.
Then both rules and script are using variable and value operations as float because it would be too much to managed typed variables.
So when you use 'epoch' it reads the 32 bit counter and cast it to float

I wasn't aware of that define @gemu2015
Is that script specific or would also work with rules?

**Reply by gemu2015** [reply]
currently only for scripts

**Reply by gemu2015** [reply]
best would be if we had a permanent settable system variable with epoch offset

**Reply by sfromis** [reply]
While rules and scripting share the limitation of doing arithmetic in single-precision floating point, in rules you can still get the full epoch value using `%utctime%` (effectively as a string) and a circuitous way of getting 31-bit integer add/subtract is to go via a counter. I don't know if similar tricks could be workable with scripting.

**Comment by gemu2015** [comment]
linked with eagle.flash.4m3m.ld you have 3Mb file system and plenty of ram (with web server)
Größe Programm	519 KB
Flash frei	504 KB
Freier Arbeitsspeicher	24.8 KB

**Reply by pkkrusty** [reply]
Yeah I noticed the 4M3M option, and have switched to using that.

**Reply by gemu2015** [reply]
@pkkrusty 
since you are struggling with precision of float, my latest dev has the option to change number format to double.
#define TS_FLOAT double
will use more RAM and be slower but full precision also on bit mangling

**Reply by pkkrusty** [reply]
Linker file desciption, example: `; Build variant ESP32 8M Flash, Tasmota 2944k Code/OTA, 2112k LITTLEFS (UFS)` seems to indicate that UFS is the same thing as LittleFS? Thus `#define USE_UFILESYS` enables a file sysyem based on LittleFS?

I ask because I'm concerned about flash writes/longevity, and trying to figure out block size for Tasmota implementation of LittleFS, so I can minimize writes, but also minimize data loss on a power cut. If I'm only writing12 bytes at a time, and block size is 104 bytes, it's not a huge deal, but if block size is 4096 bytes, I'll kill the flash real quick. 

Not even sure if I'm off-base in my thinking about file system writes in the first place...but want to intelligently consider how I'm writing to flash instead of assuming it will all be ok.

**Comment by Jason2866** [comment]
The block size is 0x1000. A smaller size makes no sense since it is the smallest size writing to flash area is possible. Anyways LittleFS takes care of wear leveling.

**Reply by barbudor** [reply]
That's the default of file systems on flash

When managing direct flash write you can tweak and take advantage that you only write 0s to flash (erase make it all 1s) so you can write multiple times to the same block as far as your unused data are all 1s

**Comment by pkkrusty** [comment]
@gemu2015 If I `#undef USE_TIMERS` can I still build a firmware with `#define USE_SCRIPT_TIMER` and be able to call `tsN(ms)` and declare timer variables `t:timer1=30`?

Declaring a t:timervar happens in >D section, and on boot the countdown begins. How to reset or re-use that timer? Just re-declare it's value in the script and countdown begins again? 

Documentation doesn't say what happens if timer variable is given value of -1 like in scripting cookbook. Does that stop the timer? 

Ticker callback section >tiN will trigger only on tsN(ms)?

**Reply by gemu2015** [reply]
#undef USE_TIMERS relates to Tasmota timers, not to script timers.
this definition is only needed to use sunrise ans sunset variables in scripter.
USE_SCRIPT_TIMER relates to so called tickers from Arduino.
you start a ticker e.g. ts1(1000) and the callback >t1 is called after 1000 ms. you must restart the ticker in the callback to get a repeated call.
the standard script timers do not need a #define and are always enabled.
t:timer=30
these are countdown timers.  if you set them to -1 they are disabled. if you set it > 0 it counts down.
so you check in >F or >S if the timer has reached zero
if timer==0 {
timer=-1
;or restart
timer = 30
}

**Reply by pkkrusty** [reply]
I tried using res=fsi(1) to return remaining space available in littlefs partition (I used 4M3M linker file) but get a "var not found" error. Does this function only work with SD Card?

**Reply by gemu2015** [reply]
#define USE_SCRIPT_FATFS_EXT
is needed for extended file system calls

btw as you use my fork. a user reported a severe error that caused numeric values to become occasionally zero.  i fixed this today. you should probably try my fork from today.
